### PR TITLE
Automatically merge green backport PRs and green approved PRs

### LINKED ISSUE #40325: Automatically merge the best pull requests
If a pull request is approved and has no modifications after it is approved and all the checks are green and there are more than 100 checks, merge automatically.

https://github.com/ClickHouse/ClickHouse/blob/be29057de1835f6f4a17e03a422b45b81efe6833/docs/ru/whats-new/extended-roadmap.md#735-%D0%BD%D0%B0%D1%87%D0%B0%D0%BB%D1%8C%D0%BD%D1%8B%D0%B5-%D0%BF%D1%80%D0%B0%D0%B2%D0%B8%D0%BB%D0%B0-%D0%B4%D0%BB%D1%8F-%D0%B0%D0%B2%D1%82%D0%BE-merge-nachalnye-pravila-dlia-avto-merge

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Automatically merge green backport PRs and green approved PRs, fixes #40325

### CODE REVIEW

**Comment by Felixoid** [comment]
```
> python merge_pr.py --check-approved --pr 41110
2022-09-08 20:36:48,885 Going to process PR #41110 in repo ClickHouse/ClickHouse
2022-09-08 20:36:48,892 Found credentials in shared credentials file: ~/.aws/credentials
2022-09-08 20:36:52,084 Checking the PR for approvals
2022-09-08 20:36:52,321 There aren't reviews for PR #41110
2022-09-08 20:36:52,321 We don't merge the PR
```

executing it locally. Would somebody approve it so I can test the script?

File: tests/ci/merge_pr.py
Comment (azat) [inline tests/ci/merge_pr.py:60]: Right now anybody can approve PR, for example me, but I don't have write access.
So `review_per_user` should be checked for write access, right?
Or maybe it works like this already and I'm missing something?

Comment (azat) [reply tests/ci/merge_pr.py:60]: Also it seems that it does not check that there were no changes after review? (like issue describes)
But maybe project had been reconfigured and now approve will be removed after new changes.

Actually most of the time subsequent changes may indeed fix failed checks correctly, and I'm pretty sure that there will be the case when I would wish for such behavior, but tricky cases are possible...

Comment (Felixoid) [reply tests/ci/merge_pr.py:60]: can you really approve a PR? it sounds suspicious

The second comment makes perfect sense, and `is_approved` should check for changes after approval

**Review by azat — APPROVED** [review]
Testing approve

File: tests/ci/merge_pr.py
Comment (azat) [reply tests/ci/merge_pr.py:60]: >can you really approve a PR?

As you can see below

>[azat](https://github.com/azat) approved these changes [2 days ago](https://github.com/ClickHouse/ClickHouse/pull/41110#pullrequestreview-1103124112)

And it is not only me, but anyone, since default policy is to allow anyone - https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-pull-request-reviews-in-your-repository

But indeed it has different color though.

**Comment by Felixoid** [comment]
Oh dear, the logic of taking the latest review per user isn't a good idea

**Comment by Felixoid** [comment]
```
In [109]: repo.get_pull(41110).head.repo.pushed_at
Out[109]: datetime.datetime(2022, 9, 12, 13, 35, 11)

In [110]: repo.get_pull(41110).head.repo.pushed_at
Out[110]: datetime.datetime(2022, 9, 12, 13, 35, 19)

In [111]: repo.get_pull(41110).head.repo.pushed_at
Out[111]: datetime.datetime(2022, 9, 12, 13, 35, 19)

In [112]: repo.get_pull(41110).head.repo.pushed_at
Out[112]: datetime.datetime(2022, 9, 12, 13, 35, 52)
```

Hmm, doesn't look like `pushed_at` works

**Comment by Felixoid** [comment]
I think it's OK to have approval from collaborators as a marker that the PR is generally ok.

**Comment by Felixoid** [comment]
Oh no, the PR would be merged on Finish in any case, even if there are failed statuses. It needs another check

**Comment by azat** [comment]
Maybe it also worth to check:
- that the PR is not in a draft state? (just in case)
- that there are no unresolved comments

Actually right now github provide everything similar:
- Require approvals
- Dismiss stale pull request approvals when new commits are pushed
- Require review from Code Owners
- Require status checks to pass before merging
- Require conversation resolution before merging

The only problem is that you have to have write access to merge PR, so this is great, but will not work.

Actually I found [auto-merge](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-auto-merge-for-pull-requests-in-your-repository).

And AFAIU anybody can select it, and this looks better then simply auto merge, maybe an author will find more problems with the patch.

P.S. llvm uses [different policy](https://llvm.org/docs/DeveloperPolicy.html#id17)

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
We have calculated `self.approved_at` without distinguishing the `author_association`. So, someone outside of the organization can approve after commits have been made after another legitimate approval.

File: tests/ci/merge_pr.py
Comment (alexey-milovidov) [inline tests/ci/merge_pr.py:76]: Here ^

**Comment by alexey-milovidov** [comment]
We can limit the recognized approvals to the core team.

**Comment by Felixoid** [comment]
@azat the GH's approach is pretty dumb and straightforward.

- First, we don't want to force approvals, and either they are strictly required, or aren't at all
- The status checks are as well very, very unflexible. Either it completely blocks the merge, or it is just ignored completely.

These two don't allow us to consider the [solution](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request), it just won't work for us

**Comment by Felixoid** [comment]
I think it's ready for review.

File: tests/ci/merge_pr.py
Comment (azat) [inline tests/ci/merge_pr.py:169]: Shouldn't it check that the checks are green?

File: tests/ci/merge_pr.py
Comment (azat) [inline tests/ci/merge_pr.py:165]: Maybe also check the subject, for some patterns?
- `WIP` - https://github.com/ClickHouse/ClickHouse/pulls?q=is%3Aopen+is%3Apr+wip
- `Do not merge`
- `Draft`

File: tests/ci/merge_pr.py
Comment (Felixoid) [reply tests/ci/merge_pr.py:169]: It does. The `check_green` is `True` by default. Maybe I get the question wrong?

File: tests/ci/merge_pr.py
Comment (Felixoid) [reply tests/ci/merge_pr.py:165]: I agree that WIP can be considered. Everything else is relatively complex to realize. Besides, both can be a part of the valid title.

I'd make the logic as simple as possible. If one don't won't to be merged - let him use `draft` PR

File: tests/ci/merge_pr.py
Comment (azat) [reply tests/ci/merge_pr.py:170]: No, I misread, thanks!

File: tests/ci/merge_pr.py
Comment (alexey-milovidov) [inline tests/ci/merge_pr.py:114]: Wrong help message.

File: tests/ci/merge_pr.py
Comment (alexey-milovidov) [inline tests/ci/merge_pr.py:170]: It should also check that the "Run Check" exists; otherwise, it's possible to have a few green checks and no other checks for a short period of time after the change.

Comment (alexey-milovidov) [reply tests/ci/merge_pr.py:170]: For simplicity, you can use "there are > 100 checks" as the condition.

Comment (Felixoid) [reply tests/ci/merge_pr.py:170]: `do not test`?

File: tests/ci/merge_pr.py
Comment (Felixoid) [reply tests/ci/merge_pr.py:114]: sure, copy-paste issue

File: tests/ci/merge_pr.py
Comment (Felixoid) [reply tests/ci/merge_pr.py:170]: Actually, both suggestions are rather bad.

We don't have `Run Check` in backports. Because it's set in CheckLabels. And both backports and docs PRs have much fewer checks than 100. `Mergeable Check` is the one that can be used :thinking: 

> otherwise, it's possible to have a few green checks and no other checks for a short period of time after the change.

I don't get your concern. The script is supposed to be launched at the very last step. Sure, it's a bad idea to launch it every step. What do you mean particularly?

Comment (alexey-milovidov) [reply tests/ci/merge_pr.py:175]: > The script is supposed to be launched at the very last step.

Didn't know this. Then everything is ok.

**Review by alexey-milovidov — APPROVED** [review]

File: tests/ci/merge_pr.py
Comment (Felixoid) [reply tests/ci/merge_pr.py:175]: Ok. Just to clarify, it's executed on the FinishCheck of [pull requests](https://github.com/ClickHouse/ClickHouse/blob/45c24dd91507b7032ba1d390c4fe33d8c29791c5/.github/workflows/pull_request.yml#L4391), [backport PRs](https://github.com/ClickHouse/ClickHouse/blob/45c24dd91507b7032ba1d390c4fe33d8c29791c5/.github/workflows/backport_branches.yml#L686) and [docs PRs](https://github.com/ClickHouse/ClickHouse/blob/45c24dd91507b7032ba1d390c4fe33d8c29791c5/.github/workflows/docs_check.yml#L172)

**Comment by Felixoid** [comment]
Ah, come on! https://github.com/ClickHouse/ClickHouse/actions/runs/3896203724/jobs/6652467766#step:6:41

**Comment by Felixoid** [comment]
[It's alive!](https://github.com/ClickHouse/ClickHouse/actions/runs/3901838162/jobs/6673400372#step:4:34)

Although, the [log](https://github.com/ClickHouse/ClickHouse/actions/runs/3901838162/jobs/6673400372#step:4:33) is a bit confusing. It's better to change it [here](https://github.com/ClickHouse/ClickHouse/pull/41110/files#diff-06dac4070b583f70ec75adc842c86ac3c6b97cb32e31f5606ac0b314ce2116cdR76-R80) to the `Approves from the core team`

**Review by tavplubix — APPROVED** [review]

**Pull request merged**
